pkg/hive/hive.go

```diff
@@ -0,0 +1,275 @@
+}
+
+// SaturationTracker tracks weather the saturation has changed
+type SaturationTracker interface {
+	Subscribe() (c <-chan struct{}, unsubscribe func())
+	Depth() (uint8, error)
+}
+
+type Service struct {
+	streamer          p2p.Streamer
+	logger            logging.Logger
+	addressBook       AddressBook
+	saturationTracker SaturationTracker
+	quit              chan struct{}
+}
+
+type Options struct {
+	Streamer          p2p.Streamer
+	Logger            logging.Logger
+	PeerTracker       AddressBook
+	SaturationTracker SaturationTracker
+}
+
+func New(o Options) *Service {
+	return &Service{
+		streamer:          o.Streamer,
+		logger:            o.Logger,
+		saturationTracker: o.SaturationTracker,
+		addressBook:       o.PeerTracker,
+		quit:              make(chan struct{}),
+	}
+}
+
+func (s *Service) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name: protocolName,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    initStreamName,
+				Version: initStreamVersion,
+				Handler: s.InitHandler,
+			},
+			{
+				Name:    subscribeStreamName,
+				Version: subscribeStreamVersion,
+				Handler: s.SubscribeHandler,
+			},
+			{
+				Name:    peersStreamName,
+				Version: peersStreamVersion,
+				Handler: s.PeersHandler,
+			},
+		},
+	}
+}
+
+// todo:
+// Hive is a skeleton of a hive function that should be called when the general hive service is started
+// this is an attempt to run these loops outside of the peer level, to avoid having them for every peer
+func (s *Service) Start() {
+	saturationChangeSignal, unsubSC := s.saturationTracker.Subscribe()
+	defer unsubSC()
+
+	ctx := context.Background()
+
+	// todo: maybe refactor and listen separately on these events, this is just a skeleton for now
+	for {
+		select {
+		// todo: handle channel close and similar stuff, depending on the implementation of publishers
+		case <-saturationChangeSignal:
+			depth, err := s.saturationTracker.Depth()
+			if err != nil {
+				// todo: handle err
+				return
+			}
+			if err := s.notifyDepthChanged(ctx, depth); err != nil {
+				// todo: handle err
+				return
+			}
+
+		// todo: cancel, quit, etc...
+		case <-s.quit:
+			return
+		}
+	}
+}
+
+// Init is called when the new peer is being initialized.
+// This should happen after overlay handshake is finished.
+func (s *Service) Init(ctx context.Context, peer p2p.Peer) error {
+	depth, err := s.saturationTracker.Depth()
+	if err != nil {
+		return err
+	}
+
+	stream, err := s.streamer.NewStream(ctx, peer.Address, protocolName, initStreamName, initStreamVersion)
+	if err != nil {
+		return fmt.Errorf("new stream: %w", err)
+	}
+
+	w, r := protobuf.NewWriterAndReader(stream)
+	if err := w.WriteMsgWithTimeout(messageTimeout, &pb.Subscribe{
+		Depth:         uint32(depth),
+		PeersResponse: true,
+	}); err != nil {
+		return fmt.Errorf("write message: %w", err)
+	}
+
+	var peers pb.Peers
+	if err := r.ReadMsgWithTimeout(messageTimeout, &peers); err != nil {
```

Comment:
What is your suggestion to do instead?

Reply:
well i am wondering - this is an application layer timeout. if it is to measure performance in any way it can easily be integrated into the `ping pong` protocol that will just in effect de-prioritise/disconnect peers that take a long time to RTT. I am just a bit weary of introducing timeouts into different modules since I feel these are mostly ballpark figures and that every developer might see them in a very different way (and as of such i would like to avoid these except for the cases that we can actually measure, quantify and justify the timeout value)

Reply:
Yes, I see. The timeout here is mainly to protect from peers that are malicious or suddenly get very slow. In general, limits are required for any publicly exposed service in order to protect it, and we can talk what is acceptable duration for that purpose. The idea of using RTT to prioritise peers is great, and we have talked briefly about it as the optimisation for the future.

Reply:
One more thing related to this is validity of the protocol, IMO. For example, in handshake we wanted for this order of msgs to happen. So, in order to handle malicious or broken node, timeout can solve this problem. For example, if we are waiting for response to finish the protocol, it's good to timeout not to wait inifinte amount of tame, IMO. So, in this sense, it makes sense to be flexible per protocol. On the general peer connect/disconnect and prioritisation, I think that RTT makes total sense IMO.